usr.bin/cvs/cvs.1
.\"	$OpenBSD: cvs.1,v 1.128 2015/12/24 16:54:37 mmcc Exp $
.\"
.\" Copyright (c) 2004 Jean-Francois Brousseau <jfb@openbsd.org>
.\" Copyright (c) 2004-2008 Xavier Santolaria <xsa@openbsd.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
.\" THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\" OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: December 24 2015 $
.Dt CVS 1
.Os
.Sh NAME
.Nm cvs
.Nd OpenCVS Concurrent Versioning System
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl flnQqRrtVvw
.Op Fl d Ar root
.Op Fl e Ar editor
.Xo
.Oo Fl s
.Ar var Ns = Ns Ar val Oc
.Xc
.Op Fl T Ar tmpdir
.Op Fl z Ar level
.Ar command ...
.Ek
.Sh DESCRIPTION
The
.Nm
program acts as both client and server for the use of and administration of
a CVS source repository.
CVS is used to maintain version information on files that are kept in a
repository.
Although it is more commonly used to track changes in source code, there
are no real limitations to the type of files that can be stored in a
repository.
For a general introduction to CVS, see
.Xr cvsintro 7 .
.Pp
.Nm
reads its startup configuration file,
.Pa .cvsrc ,
from the home directory of the user who invoked it.
This file is used to specify implicit options passed to
.Nm
or one of its commands whenever it is invoked.
The defaults in the configuration file can be overridden with the
.Fl f
option (see below).
See
.Xr cvs 5
for further information.
.Pp
.Nm
also supports
keyword substitution \(en
see the
.Xr rcs 1
man page for more information.
.Pp
The following options are supported:
.Bl -tag -width Ds
.It Fl d Ar root
Use
.Ar root
as the path to the root directory of the CVS repository.
The value must specify an absolute path.
.It Fl e Ar editor
Use the program
.Ar editor
whenever editing log information.
This option overrides the environment variables CVSEDITOR, VISUAL, and EDITOR.
.It Fl f
Do not read the user's configuration file on startup.
.It Fl l
Suppress logging of history information.
.It Fl n
Dry-run mode.
Show which files will be used by the command issued
without really running it.
.It Fl Q
Be extra quiet.
Only error messages will be displayed.
.It Fl q
Be quiet about reporting.
.It Fl R
Permit checkout from a read-only repository.
Implies
.Fl l .
See also
.Ev CVSREADONLYFS ,
below.
.It Fl r
Extract files in read-only mode.
.It Fl s Ar var Ns = Ns Ar val
Set the value of the internal variable
.Ar var
to the string
.Ar val .
.It Fl T Ar tmpdir
Set the value of the directory where temporary files are to be created.
The default is set to
.Pa /tmp .
This option overrides the
.Ev TMPDIR
environment variable.
.It Fl t
Trace program execution.
.It Fl V
Verbose mode.
All messages will be displayed.
This is the default.
.Fl V
and
.Fl Q
are mutually exclusive.
If both are specified,
.Fl Q
takes precedence.
.It Fl v
Display version information and exit.
.It Fl w
Extract new files in read-write mode.
Overrides the setting of the
.Ev CVSREAD
environment variable.
This is the default unless
.Ev CVSREAD
is set or the
.Fl r
option is specified.
.It Fl z Ar level
Specify the compression level to
.Xr gzip 1
when transferring files.
The compression level ranges from 1 to 9,
with 1 being the fastest,
and 9 providing the best level of compression.
The default is 6.
.El
.Sh COMMANDS
.Nm
supports the following commands:
add,
admin,
annotate,
checkout,
commit,
diff,
edit,
editors,
export,
history,
import,
init,
kserver,
log,
rannotate,
rdiff,
release,
remove,
rlog,
rtag,
server,
status,
tag,
unedit,
update,
version,
watch,
watchers.
The commands are fully explained in this section.
.Pp
Files may be selected by
.Em revision
or, where no revision is specified,
the latest revision of the default branch is used.
Revisions are specified either by using the
.Fl r
option or
by appending the revision number to any option that supports it.
.Pp
.Nm
supports the notion of
.Em state .
The state is an arbitrary string of characters used to describe a file
(or a specific revision of a file).
States can be set or changed using the
.Fl s
option, for CVS tools which support it.
The state of a file/revision can be modified without having to
.Ic commit
a new file/revision.
The default state is
.Sq Exp
(Experimental).
Examples of states could be
.Sq Dev ,
.Sq Reviewed ,
or
.Sq Stab .
.Ss add
Before a file is known to
.Nm ,
it must be added to the repository using this command.
Adding a file does not actually publish the contents of the
file: the
.Ic commit
command must also be used to publish it into the repository,
and thus let others access the file.
.Pp
Note: since directories have no versioning system, it is sufficient
to add them with the
.Ic add
command alone; the
.Ic commit
command is not necessary.
.Bd -literal -offset indent
usage: cvs add [-k mode] [-m msg] file ...
.Ed
.Pp
The
.Ic add
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl k Ar mode
Specify the keyword substitution mode.
.It Fl m Ar msg
Attach log message
.Ar msg .
By default, no log message is required.
.El
.Pp
Aliases:
.Ic ad ,
.Ic new .
.Ss admin
The
.Ic admin
command is used to directly modify the RCS files.
.Bd -literal -offset indent
usage: cvs admin [-Iq] [-b branch] [-k mode] [-m rev:msg]
                 [-N tag[:rev]] [-n tag[:rev]] [-o rev]
                 [-s state[:rev]] [-t file | str]
.Ed
.Pp
The
.Ic admin
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl b Ar branch
Set the default branch to
.Ar branch .
.It Fl I
Command is interactive.
.It Fl k Ar mode
Specify the keyword substitution mode.
.It Fl m Ar rev : Ns Ar msg
Change the log message of a revision.
.It Xo Fl N
.Ar tag Ns Op : Ns Ar rev
.Xc
Same as
.Fl n ,
but override tag if it already exists.
.It Xo Fl n
.Ar tag Ns Op : Ns Ar rev
.Xc
Associate the
.Ar tag
with the
.Ar rev
or the branch given as argument.
If the revision or the branch is not specified, the tag is deleted.
The
.Sq \&:
character means the association of the tag and the latest revision of
the default branch.
A branch number ending with the
.Sq \&.
character means the current latest revision in the branch.
This option is functionally the same as the
.Ic rtag
command, but it avoids the check of the tags done with the
.Pa CVSROOT/taginfo
file.
.It Fl o Ar rev
Delete one or more revisions.
The specifications of the values or revisions are as follows:
.Bl -tag -width Ds
.It rev
Specific revision.
.It rev1:rev2
Delete all revisions of a branch between
.Ar rev1
and
.Ar rev2 .
.It rev1::rev2
Delete all revisions of a branch between
.Ar rev1
and
.Ar rev2
without deleting revisions
.Ar rev1
and
.Ar rev2 .
.It :rev
Delete all revisions of the branch until revision
.Ar rev .
.It rev:
Delete all revisions of the branch from revision
.Ar rev
until the last revision of the branch.
.El
.It Fl q
Quiet mode.
.It Xo Fl s
.Ar state Ns Op : Ns Ar rev
.Xc
Change state of a revision.
.It Fl t Ar file \*(Ba Ar str
Change the descriptive text.
The descriptive text is taken from the
.Ar file
specified as argument or from the string
.Ar str
given as argument if it is preceded by the
.Sq -
character.
If no argument is used, the descriptive text is taken from standard input.
.El
.Pp
Aliases:
.Ic adm ,
.Ic rcs .
.Ss annotate
For each line of any files specified, show information about its
last revision.
The information given is the last revision when a modification occurred,
the author's name, and the date of the revision.
.Bd -literal -offset indent
usage: cvs annotate [-flR] [-D date | -r rev] [file ...]
.Ed
.Pp
The
.Ic annotate
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl D Ar date
Show the annotations as of the latest revision no later than
.Ar date .
.It Fl f
Force the use of the head revision if the specified
tag or date is not found.
This can be used in combination with
.Fl D
or
.Fl r
to ensure that there is some output from the
.Ic annotate
command, even if only to show Revision 1.1 of the file.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Show annotations as of revision
.Ar rev
(can be a revision number or a tag).
.El
.Pp
Aliases:
.Ic ann ,
.Ic blame .
.Ss checkout
The
.Ic checkout
command is used to create a local copy of one or more modules present on the
target CVS repository.
.Bd -literal -offset indent
usage: cvs checkout [-AcflNnPpRs] [-d dir] [-j rev] [-k mode]
                    -D date | -r rev module ...
.Ed
.Pp
The
.Ic checkout
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl A
Reset any sticky tags, dates, or keyword substitution modes that
have been set on the tree.
.It Fl c
Display the list of available modules.
.It Fl D Ar date
Check out as of the latest revision no later than
.Ar date
(implies
.Fl P )
(is sticky).
.It Fl d Ar dir
Check out in directory
.Ar dir
instead of the directory bearing the same name as the
.Ar module .
.It Fl f
Force the use of the head revision if the specified
tag or date is not found.
.It Fl j Ar rev
Merge in changes made between current revision and
.Ar rev .
If two
.Fl j
options are specified, only merge the differences between the two
revisions of the branch.
This allows successive merges without having to resolve
already resolved conflicts again.
.It Fl k Ar mode
Specify the keyword substitution mode (is sticky).
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl N
If used in conjunction with the
.Fl d
option, files are placed in local directory
.Ar module ,
located in directory
.Ar dir .
.It Fl n
Do not execute programs listed in the
.Pa CVSROOT/modules
file.
.It Fl P
Prune empty directories.
.It Fl p
Check out files to standard output (avoids stickiness).
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Check out from a particular revision or branch (implies
.Fl P )
(is sticky).
.It Fl s
Like
.Fl c ,
but include module status.
.El
.Pp
Aliases:
.Ic co ,
.Ic get .
.Ss commit
The
.Ic commit
command is used to send local changes back to the server and update the
repository's information to reflect the changes.
.Bd -literal -offset indent
usage: cvs commit [-flnR] [-F logfile | -m msg] [-r rev] [file ...]
.Ed
.Pp
The
.Ic commit
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl F Ar logfile
Specify a
.Ar file
which contains the log message.
.It Fl f
Force a file to be committed, even though it is unchanged.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl m Ar msg
Specify a log message on the command line (suppresses the editor invocation).
.It Fl n
Do not execute programs listed in the
.Pa CVSROOT/modules
file.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Fl r Ar rev
Commit to a particular symbolic or numerical revision.
.El
.Pp
Aliases:
.Ic ci ,
.Ic com .
.Ss diff
The
.Ic diff
command is very similar to the
.Xr diff 1
program, except that the differential comparisons that it generates are
between local or remote revisions of files stored in the CVS repository.
.Bd -literal -offset indent
usage: cvs diff [-abcdilNnpRuw]
                [[-D date1 | -r rev1] [-D date2 | -r rev2]]
                [-k mode] [file ...]
.Ed
.Pp
The
.Ic diff
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl a
Treat all files as ASCII text.
See
.Xr diff 1
for more information.
.It Fl b
Causes trailing blanks (spaces and tabs) to be ignored, and other
strings of blanks to compare equal.
.It Fl c
Produces a diff with three lines of context.
See
.Xr diff 1
for more information.
.It Xo Fl D Ar date1
.Op Fl D Ar date2
.Xc
Differences between the revision at
.Ar date1
and the working copy or
.Ar date1
and
.Ar date2
(if specified).
.It Fl d
Try very hard to produce a diff as small as possible.
See
.Xr diff 1
for more information.
.It Fl i
Ignore the case of letters.
For example,
.Sq A
will compare equal to
.Sq a .
.It Fl k Ar mode
Specify the keyword substitution mode.
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl N
Include added or removed files.
.It Fl n
Produces a diff in the same format as that used by
.Xr rcsdiff 1 ,
with a count of changed lines on each insert or delete command.
.It Fl p
With unified and context diffs, show with each change the first
40 characters of the last line before the context beginning with
a letter, an underscore or a dollar sign.
See
.Xr diff 1
for more information.
.It Fl R
Enable recursive behaviour.
This is the default.
.It Xo Fl r Ar rev1
.Op Fl r Ar rev2
.Xc
Differences between revision
.Ar rev1
and the working copy or
.Ar rev1
and
.Ar rev2
(if specified).
.It Fl t
Will expand tabs in output lines.
Normal or
.Fl c
output adds character(s) to the front of each line which may screw up
the indentation of the original source lines and make the output listing
difficult to interpret.
This option will preserve the original source's indentation.
.It Fl u
Produces a unified diff with three lines of context.
See
.Xr diff 1
for more information.
.It Fl w
Is similar to
.Fl b
but causes whitespace (blanks and tabs) to be totally ignored.
For example,
.Dq if (\ \&a == b \&)
will compare equal to
.Dq if(a==b) .
.El
.Pp
Aliases:
.Ic di ,
.Ic dif .
.Ss edit
The
.Ic edit
command is used to make a file that is being watched
(and therefore read-only)
readable and writable and to inform others that it is in the
process of being changed.
Notifications terminate when the
.Ic commit
command is issued.
Editing rights on the file can be given up using the
.Ic unedit
command, which terminates the temporary notifications.
.Bd -literal -offset indent
usage: cvs edit [-lR] [-a action] [file ...]
.Ed
.Pp
The
.Ic edit
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl a Ar action
Specify the temporary notification wanted:
.Pp
.Bl -tag -width Ds -compact
.It Cm commit
Another user has committed changes to the file.
.It Cm edit
Another user has issued the
.Ic edit
command on the file.
.It Cm unedit
Another user has issued the
.Ic unedit
command on the file.
.It Cm all
All of the above.
.It Cm none
None of the above.
.El
.Pp
The
.Fl a
flag may appear more than once, or not at all.
If omitted, the action defaults to
.Cm all .
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.Ss editors
The
.Ic editors
command lists the users with edition rights on a file.
For that, pseudo-lock mode must be enabled (see the
.Ic watch
command).
The email address of the user editing the file, the timestamp
when the edition first started, the host from where the edition
has been requested and the path to the edited file are listed.
.Bd -literal -offset indent
usage: cvs editors [-lR] [file ...]
.Ed
.Pp
The
.Ic editors
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl l
Limit the scope of the search to the local directory
only and disable recursive behaviour.
.It Fl R
Enable recursive behaviour.
This is the default.
.El
.Ss export
The
.Ic export
command extracts a copy of
.Ar module
without including the directories used for management by
.Nm .
This eases production of a software release.
A date or a revision must be specified for the command to be valid,
which ensures that later extractions can be reproduced with the same
options as the release.
.Pp
The checked out module's files will be placed in a directory
bearing the same name as the checked out module, by default.
.Bd -literal -offset indent
usage: cvs export [-flNnR] [-d dir] [-k mode]
                  -D date | -r rev module ...
.Ed
.Pp
The
.Ic export
command takes the following options:
.Bl -tag -width Ds -offset 3n
.It Fl D Ar date
Export as of the latest revision no later than
.Ar date .
.It Fl d Ar dir
Export in directory
.Ar dir

usr.bin/cvs/cvs.c
/*	$OpenBSD: cvs.c,v 1.157 2015/11/19 17:44:01 deraadt Exp $	*/
/*
 * Copyright (c) 2006, 2007 Joris Vink <joris@openbsd.org>
 * Copyright (c) 2004 Jean-Francois Brousseau <jfb@openbsd.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 * THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/stat.h>

#include <ctype.h>
#include <errno.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <err.h>

#include "cvs.h"
#include "remote.h"
#include "hash.h"

extern char *__progname;

/* verbosity level: 0 = really quiet, 1 = quiet, 2 = verbose */
int verbosity = 2;

/* compression level used with zlib, 0 meaning no compression taking place */
int	cvs_compress = 0;
int	cvs_readrc = 1;		/* read .cvsrc on startup */
int	cvs_trace = 0;
int	cvs_nolog = 0;
int	cvs_readonly = 0;
int	cvs_readonlyfs = 0;
int	cvs_nocase = 0;	/* set to 1 to disable filename case sensitivity */
int	cvs_noexec = 0;	/* set to 1 to disable disk operations (-n option) */
int	cvs_cmdop;
int	cvs_umask = CVS_UMASK_DEFAULT;
int	cvs_server_active = 0;

char	*cvs_tagname = NULL;
char	*cvs_defargs;		/* default global arguments from .cvsrc */
char	*cvs_rootstr;
char	*cvs_rsh = CVS_RSH_DEFAULT;
char	*cvs_editor = CVS_EDITOR_DEFAULT;
char	*cvs_homedir = NULL;
char	*cvs_tmpdir = CVS_TMPDIR_DEFAULT;

struct cvsroot *current_cvsroot = NULL;
struct cvs_cmd *cmdp;			/* struct of command we are running */

int		cvs_getopt(int, char **);
__dead void	usage(void);
static void	cvs_read_rcfile(void);

struct wklhead temp_files;

void sighandler(int);
volatile sig_atomic_t cvs_quit = 0;
volatile sig_atomic_t sig_received = 0;

extern CVSENTRIES *current_list;

struct hash_table created_directories;
struct hash_table created_cvs_directories;

void
sighandler(int sig)
{
	sig_received = sig;

	switch (sig) {
	case SIGINT:
	case SIGTERM:
	case SIGPIPE:
		cvs_quit = 1;
		break;
	default:
		break;
	}
}

void
cvs_cleanup(void)
{
	cvs_log(LP_TRACE, "cvs_cleanup: removing locks");
	worklist_run(&repo_locks, worklist_unlink);

	cvs_log(LP_TRACE, "cvs_cleanup: removing temp files");
	worklist_run(&temp_files, worklist_unlink);

	if (cvs_server_path != NULL) {
		if (cvs_rmdir(cvs_server_path) == -1)
			cvs_log(LP_ERR,
			    "warning: failed to remove server directory: %s",
			    cvs_server_path);
		free(cvs_server_path);
		cvs_server_path = NULL;
	}

	if (current_list != NULL)
		cvs_ent_close(current_list, ENT_SYNC);
}

__dead void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: %s [-flnQqRrtvw] [-d root] [-e editor] [-s var=val]\n"
	    "           [-T tmpdir] [-z level] command ...\n", __progname);
	exit(1);
}

int
cvs_build_cmd(char ***cmd_argv, char **argv, int argc)
{
	int cmd_argc, i, cur;
	char *cp, *linebuf, *lp;

	if (cmdp->cmd_defargs == NULL) {
		*cmd_argv = argv;
		return argc;
	}

	cur = argc + 2;
	cmd_argc = 0;
	*cmd_argv = xcalloc(cur, sizeof(char *));
	(*cmd_argv)[cmd_argc++] = argv[0];

	linebuf = xstrdup(cmdp->cmd_defargs);
	for (lp = linebuf; lp != NULL;) {
		cp = strsep(&lp, " \t\b\f\n\r\t\v");
		if (cp == NULL)
			break;
		if (*cp == '\0')
			continue;

		if (cmd_argc == cur) {
			cur += 8;
			*cmd_argv = xreallocarray(*cmd_argv, cur,
			    sizeof(char *));
		}

		(*cmd_argv)[cmd_argc++] = cp;
	}

	if (cmd_argc + argc > cur) {
		cur = cmd_argc + argc + 1;
		*cmd_argv = xreallocarray(*cmd_argv, cur,
		    sizeof(char *));
        }

	for (i = 1; i < argc; i++)
		(*cmd_argv)[cmd_argc++] = argv[i];

	(*cmd_argv)[cmd_argc] = NULL;

	return cmd_argc;
}

int
main(int argc, char **argv)
{
	char *envstr, **cmd_argv, **targv;
	int i, ret, cmd_argc;
	struct passwd *pw;
	struct stat st;
	char fpath[PATH_MAX];

	if (pledge("stdio rpath wpath cpath fattr proc exec", NULL) == -1)
		err(1, "pledge");

	tzset();

	TAILQ_INIT(&cvs_variables);
	SLIST_INIT(&repo_locks);
	SLIST_INIT(&temp_files);

	hash_table_init(&created_directories, 100);
	hash_table_init(&created_cvs_directories, 100);

	/* check environment so command-line options override it */
	if ((envstr = getenv("CVS_RSH")) != NULL)
		cvs_rsh = envstr;

	if (((envstr = getenv("CVSEDITOR")) != NULL) ||
	    ((envstr = getenv("VISUAL")) != NULL) ||
	    ((envstr = getenv("EDITOR")) != NULL))
		cvs_editor = envstr;

	if ((envstr = getenv("CVSREAD")) != NULL)
		cvs_readonly = 1;

	if ((envstr = getenv("CVSREADONLYFS")) != NULL) {
		cvs_readonlyfs = 1;
		cvs_nolog = 1;
	}

	if ((cvs_homedir = getenv("HOME")) == NULL) {
		if ((pw = getpwuid(getuid())) != NULL)
			cvs_homedir = pw->pw_dir;
	}

	if ((envstr = getenv("TMPDIR")) != NULL)
		cvs_tmpdir = envstr;

	ret = cvs_getopt(argc, argv);

	argc -= ret;
	argv += ret;
	if (argc == 0)
		usage();

	cmdp = cvs_findcmd(argv[0]);
	if (cmdp == NULL) {
		fprintf(stderr, "Unknown command: `%s'\n\n", argv[0]);
		fprintf(stderr, "CVS commands are:\n");
		for (i = 0; cvs_cdt[i] != NULL; i++)
			fprintf(stderr, "\t%-16s%s\n",
			    cvs_cdt[i]->cmd_name, cvs_cdt[i]->cmd_descr);
		exit(1);
	}

	/*
	 * check the tmp dir, either specified through
	 * the environment variable TMPDIR, or via
	 * the global option -T <dir>
	 */
	if (stat(cvs_tmpdir, &st) == -1)
		fatal("stat failed on `%s': %s", cvs_tmpdir, strerror(errno));
	else if (!S_ISDIR(st.st_mode))
		fatal("`%s' is not valid temporary directory", cvs_tmpdir);

	if (cvs_readrc == 1 && cvs_homedir != NULL) {
		cvs_read_rcfile();

		if (cvs_defargs != NULL) {
			if ((targv = cvs_makeargv(cvs_defargs, &i)) == NULL)
				fatal("failed to load default arguments to %s",
				    __progname);

			cvs_getopt(i, targv);
			cvs_freeargv(targv, i);
			free(targv);
		}
	}

	/* setup signal handlers */
	signal(SIGTERM, sighandler);
	signal(SIGINT, sighandler);
	signal(SIGHUP, sighandler);
	signal(SIGABRT, sighandler);
	signal(SIGALRM, sighandler);
	signal(SIGPIPE, sighandler);

	cvs_cmdop = cmdp->cmd_op;

	cmd_argc = cvs_build_cmd(&cmd_argv, argv, argc);

	cvs_file_init();

	if (cvs_cmdop == CVS_OP_SERVER) {
		cmdp->cmd(cmd_argc, cmd_argv);
		cvs_cleanup();
		return (0);
	}

	cvs_umask = umask(0);
	umask(cvs_umask);

	if ((current_cvsroot = cvsroot_get(".")) == NULL) {
		cvs_log(LP_ERR,
		    "No CVSROOT specified! Please use the '-d' option");
		fatal("or set the CVSROOT environment variable.");
	}

	if (current_cvsroot->cr_method != CVS_METHOD_LOCAL) {
		cmdp->cmd(cmd_argc, cmd_argv);
		cvs_cleanup();
		return (0);
	}

	(void)xsnprintf(fpath, sizeof(fpath), "%s/%s",
	    current_cvsroot->cr_dir, CVS_PATH_ROOT);

	if (stat(fpath, &st) == -1 && cvs_cmdop != CVS_OP_INIT) {
		if (errno == ENOENT)
			fatal("repository '%s' does not exist",
			    current_cvsroot->cr_dir);
		else
			fatal("%s: %s", current_cvsroot->cr_dir,
			    strerror(errno));
	} else {
		if (!S_ISDIR(st.st_mode))
			fatal("'%s' is not a directory",
			    current_cvsroot->cr_dir);
	}

	if (cvs_cmdop != CVS_OP_INIT) {
		cvs_parse_configfile();
		cvs_parse_modules();
	}

	cmdp->cmd(cmd_argc, cmd_argv);
	cvs_cleanup();

	return (0);
}

int
cvs_getopt(int argc, char **argv)
{
	int ret;
	char *ep;
	const char *errstr;

	while ((ret = getopt(argc, argv, "b:d:e:flnQqRrs:T:tvwxz:")) != -1) {
		switch (ret) {
		case 'b':
			/*
			 * We do not care about the bin directory for RCS files
			 * as this program has no dependencies on RCS programs,
			 * so it is only here for backwards compatibility.
			 */
			cvs_log(LP_NOTICE, "the -b argument is obsolete");
			break;
		case 'd':
			cvs_rootstr = optarg;
			break;
		case 'e':
			cvs_editor = optarg;
			break;
		case 'f':
			cvs_readrc = 0;
			break;
		case 'l':
			cvs_nolog = 1;
			break;
		case 'n':
			cvs_noexec = 1;
			cvs_nolog = 1;
			break;
		case 'Q':
			verbosity = 0;
			break;
		case 'q':
			if (verbosity > 1)
				verbosity = 1;
			break;
		case 'R':
			cvs_readonlyfs = 1;
			cvs_nolog = 1;
			break;
		case 'r':
			cvs_readonly = 1;
			break;
		case 's':
			ep = strchr(optarg, '=');
			if (ep == NULL) {
				cvs_log(LP_ERR, "no = in variable assignment");
				exit(1);
			}
			*(ep++) = '\0';
			if (cvs_var_set(optarg, ep) < 0)
				exit(1);
			break;
		case 'T':
			cvs_tmpdir = optarg;
			break;
		case 't':
			cvs_trace = 1;
			break;
		case 'v':
			printf("%s\n", CVS_VERSION);
			exit(0);
			/* NOTREACHED */
		case 'w':
			cvs_readonly = 0;
			break;
		case 'x':
			/*
			 * Kerberos encryption support, kept for compatibility
			 */
			break;
		case 'z':
			cvs_compress = strtonum(optarg, 0, 9, &errstr);
			if (errstr != NULL)
				fatal("cvs_compress: %s", errstr);
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}

	ret = optind;
	optind = 1;
	optreset = 1;	/* for next call */

	return (ret);
}

/*
 * cvs_read_rcfile()
 *
 * Read the CVS `.cvsrc' file in the user's home directory.  If the file
 * exists, it should contain a list of arguments that should always be given
 * implicitly to the specified commands.
 */
static void
cvs_read_rcfile(void)
{
	char rcpath[PATH_MAX], *buf, *lbuf, *lp, *p;
	int cmd_parsed, cvs_parsed, i, linenum;
	size_t len, pos;
	struct cvs_cmd *tcmdp;
	FILE *fp;

	linenum = 0;

	i = snprintf(rcpath, PATH_MAX, "%s/%s", cvs_homedir, CVS_PATH_RC);
	if (i < 0 || i >= PATH_MAX) {
		cvs_log(LP_ERRNO, "%s", rcpath);
		return;
	}

	fp = fopen(rcpath, "r");
	if (fp == NULL) {
		if (errno != ENOENT)
			cvs_log(LP_NOTICE, "failed to open `%s': %s", rcpath,
			    strerror(errno));
		return;
	}

	cmd_parsed = cvs_parsed = 0;
	lbuf = NULL;
	while ((buf = fgetln(fp, &len)) != NULL) {
		if (buf[len - 1] == '\n') {
			buf[len - 1] = '\0';
		} else {
			lbuf = xmalloc(len + 1);
			memcpy(lbuf, buf, len);
			lbuf[len] = '\0';
			buf = lbuf;
		}

		linenum++;

		/* skip any whitespaces */
		p = buf;
		while (*p == ' ')
			p++;

		/*
		 * Allow comments.
		 * GNU cvs stops parsing a line if it encounters a \t
		 * in front of a command, stick at this behaviour for
		 * compatibility.
		 */
		if (*p == '#' || *p == '\t')
			continue;

		pos = strcspn(p, " \t");
		if (pos == strlen(p)) {
			lp = NULL;
		} else {
			lp = p + pos;
			*lp = '\0';
		}

		if (strcmp(p, "cvs") == 0 && !cvs_parsed) {
			/*
			 * Global default options.  In the case of cvs only,
			 * we keep the 'cvs' string as first argument because
			 * getopt() does not like starting at index 0 for
			 * argument processing.
			 */
			if (lp != NULL) {
				*lp = ' ';
				cvs_defargs = xstrdup(p);
			}
			cvs_parsed = 1;
		} else {
			tcmdp = cvs_findcmd(p);
			if (tcmdp == NULL && verbosity == 2)
				cvs_log(LP_NOTICE,
				    "unknown command `%s' in `%s:%d'",
				    p, rcpath, linenum);

			if (tcmdp != cmdp || cmd_parsed)
				continue;

			if (lp != NULL) {
				lp++;
				cmdp->cmd_defargs = xstrdup(lp);
			}
			cmd_parsed = 1;
		}
	}
	free(lbuf);

	if (ferror(fp)) {
		cvs_log(LP_NOTICE, "failed to read line from `%s'", rcpath);
	}

	(void)fclose(fp);
}

/*
 * cvs_var_set()
 *
 * Set the value of the variable <var> to <val>.  If there is no such variable,
 * a new entry is created, otherwise the old value is overwritten.
 * Returns 0 on success, or -1 on failure.
 */
int
cvs_var_set(const char *var, const char *val)
{
	const char *cp;
	struct cvs_var *vp;

	if (var == NULL || *var == '\0') {
		cvs_log(LP_ERR, "no variable name");
		return (-1);
	}

	/* sanity check on the name */
	for (cp = var; *cp != '\0'; cp++)
		if (!isalnum((unsigned char)*cp) && (*cp != '_')) {
			cvs_log(LP_ERR,
			    "variable name `%s' contains invalid characters",
			    var);
			return (-1);
		}

	TAILQ_FOREACH(vp, &cvs_variables, cv_link)
		if (strcmp(vp->cv_name, var) == 0)
			break;

	if (vp == NULL) {
		vp = xcalloc(1, sizeof(*vp));

		vp->cv_name = xstrdup(var);
		TAILQ_INSERT_TAIL(&cvs_variables, vp, cv_link);

	} else	/* free the previous value */
		free(vp->cv_val);

	vp->cv_val = xstrdup(val);

	return (0);
}

/*
 * cvs_var_unset()
 *
 * Remove any entry for the variable <var>.
 * Returns 0 on success, or -1 on failure.
 */
int
cvs_var_unset(const char *var)
{
	struct cvs_var *vp;

	TAILQ_FOREACH(vp, &cvs_variables, cv_link)
		if (strcmp(vp->cv_name, var) == 0) {
			TAILQ_REMOVE(&cvs_variables, vp, cv_link);
			free(vp->cv_name);
			free(vp->cv_val);
			free(vp);
			return (0);
		}

	return (-1);
}

/*
 * cvs_var_get()
 *
 * Get the value associated with the variable <var>.  Returns a pointer to the
 * value string on success, or NULL if the variable does not exist.
 */

const char *
cvs_var_get(const char *var)
{
	struct cvs_var *vp;

	TAILQ_FOREACH(vp, &cvs_variables, cv_link)
		if (strcmp(vp->cv_name, var) == 0)
			return (vp->cv_val);

	return (NULL);
}
